lib/ddtrace/contrib/rails/patcher.rb

```diff
@@ -45,17 +45,33 @@ def before_intialize(app)
         end
 
         def add_middleware(app)
-          # Add trace middleware
+          # Add trace middleware at the top of the middleware stack,
+          # to ensure we capture the complete execution time.
           app.middleware.insert_before(0, Datadog::Contrib::Rack::TraceMiddleware)
 
-          # Insert right after Rails exception handling middleware, because if it's before,
-          # it catches and swallows the error. If it's too far after, custom middleware can find itself
-          # between, and raise exceptions that don't end up getting tagged on the request properly.
-          # e.g lost stack trace.
-          app.middleware.insert_after(
-            ActionDispatch::ShowExceptions,
-            Datadog::Contrib::Rails::ExceptionMiddleware
-          )
+          # Some Rails middleware can swallow an application error, preventing
+          # the error propagation to the encompassing Rack span.
+          #
+          # We insert our own middleware right before these Rails middleware
+          # have a chance to swallow the error.
+          #
+          # Because there's more than one possible middleware that swallows errors,
+          # we insert our middleware before all the offending ones.
+          # At this point of the application lifecycle the complete middleware stack,
+          # is not yet defined, thus we can't decide which exact Rails middleware we want to
+          # insert ourselves before. We resort to inserting ourselves as many times as needed,
+          # once per offending middleware.
```

Comment:
I'm curious, where are we "inserting ourselves as many times as needed"? Is that loop-like behavior elsewhere?

Reply:
This comment was from when I thought we had to insert our middleware before either `DebugExceptions` or `ShowExceptions`, whichever came first. But the truth was that we should insert ourselves before `DebugExceptions`, as it sits closer to the top of the Rails stack. (save for Rails 3.0, which didn't have `DebugExceptions` yet).

Turns out it was simpler than I thought.